internal/node/otel.go

```
@@ -85,35 +122,106 @@ func NewOTELClient(
+	creds := grpc.WithTransportCredentials(insecure.NewCredentials())
+
+	client.grpcConn, err = grpc.NewClient(config.GRPCEndpoint, creds)
 	if err != nil {
 		return nil, fmt.Errorf("creating new grpc connection: %w", err)
 	}
 
-	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
+	// -- Tracing
+
+	client.TracerProvider, err = newTracerProvider(ctx, client.grpcConn, server)
 	if err != nil {
-		return nil, fmt.Errorf("creating a trace exporter: %w", err)
+		closeClient()
+		return nil, errors.Wrap(err, "generating a tracerProvider")
 	}
 
-	// -- TracerProvider
+	// set propagation to include traceContext and baggage (the default is no-op).
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{}))
+	otel.SetTracerProvider(client.TracerProvider)
+	client.Tracer = client.TracerProvider.Tracer(serviceName + "/tracer")
+
+	// -- Logging
+
+	// generate a logger provider
+	// LoggerProvider := global.GetLoggerProvider()
+	client.LoggerProvider, err = newLoggerProvider(ctx, client.grpcConn, server)
+	if err != nil {
+		closeClient()
+		return nil, errors.Wrap(err, "generating a tracerProvider")
+	}
+
+	global.SetLoggerProvider(client.LoggerProvider)
+	client.Logger = client.LoggerProvider.Logger(serviceName)
+
+	// -- Metrics
+
+	client.MeterProvider, err = newMeterProvider(ctx, client.grpcConn, server)
+	if err != nil {
+		closeClient()
+		return nil, errors.Wrap(err, "generating a tracerProvider")
+	}
+
+	otel.SetMeterProvider(client.MeterProvider)
+	client.Meter = client.MeterProvider.Meter(serviceName)
+
+	// Shutdown will flush any remaining spans and shut down the exporter.
+	return &client, nil
+}
+
+// newTracerProvider constructs a new tracer that manages batch exports
+// of tracing values.
+func newTracerProvider(
+	ctx context.Context,
+	conn *grpc.ClientConn,
+	server *resource.Resource,
+) (*sdkTrace.TracerProvider, error) {
+	if ctx != nil {
```

Few questions:
1. What's the reasoning behind the 5 sec timeout?
2. If the caller supplies a nil ctx, we'll fail to apply the 5 sec timeout to the request. That means request can run indefinitely. Instead, would it be better to return an error if ctx is nil?

---

1/ I think it's to avoid hanging on the connection request.  I think it's unnecessary in most cases, but also a common recommend in example code.

2/ good point.  I don't think an error is correct, but I can indeed return.

---

Sounds good!